json/src/com/jetbrains/jsonSchema/impl/JsonSchemaFileValuesIndex.java

```diff
@@ -44,15 +48,60 @@
         final FileType fileType = inputData.getFileType();
         if (fileType != JsonFileType.INSTANCE
             && fileType != Json5FileType.INSTANCE) return ContainerUtil.newHashMap();
-        PsiFile file = inputData.getPsiFile();
-        if (!(file instanceof JsonFile)) return ContainerUtil.newHashMap();
-        HashMap<String, String> map = ContainerUtil.newHashMap();
-        String schemaUrl = JsonCachedValues.fetchSchemaUrl(file);
-        map.put(JsonCachedValues.URL_CACHE_KEY, schemaUrl == null ? NULL : schemaUrl);
-        String schemaId = JsonCachedValues.fetchSchemaId(file);
-        map.put(JsonCachedValues.ID_CACHE_KEY, schemaId == null ? NULL : schemaId);
+        final HashMap<String, String> map = ContainerUtil.newHashMap();
+        final Lexer lexer = fileType == Json5FileType.INSTANCE ? new Json5Lexer() : new JsonLexer();
+        lexer.start(inputData.getContentAsText());
+
+        // We only care about properties at the root level having the form of "property" : "value".
+        int nesting = 0;
+        while (lexer.getCurrentPosition().getOffset() < lexer.getBufferEnd()) {
+          IElementType token = lexer.getTokenType();
+          // Nesting level can only change at curly braces.
+          if (token == JsonElementTypes.L_CURLY) {
+            nesting++;
+          }
+          else if (token == JsonElementTypes.R_CURLY) {
+            nesting--;
+          }
+          else if (nesting == 1 && (token == JsonElementTypes.DOUBLE_QUOTED_STRING || token == JsonElementTypes.IDENTIFIER)) {
+            // We are looking for two special properties at the root level.
+            switch (lexer.getTokenText()) {
+              case "\"$id\"":
+              case "$id":
+                captureValueIfString(lexer, map, JsonCachedValues.ID_CACHE_KEY);
+                break;
+              case "\"$schema\"":
+              case "$schema":
+                captureValueIfString(lexer, map, JsonCachedValues.URL_CACHE_KEY);
+                break;
+            }
+          }
+          lexer.advance();
```

Comment:
Looks like we can break the iteration if we've already found both properties. In most real-life cases, these properties are located on top, so skipping the remaining part should help a lot.

Reply:
Thank you! Done. It won't help in the case when none is present, though.
I've also allowed the keys to be single-quoted for compatibility with json5.

Reply:
Thanks. Yes, that's true, there's nothing to improve when none of the properties is there.
Merged the PR, thanks for your work.